### ISSUE TITLE
Proposal: Increase BEV Feature Map Resolution

### ISSUE THREAD

**Issue by riita10069** [autowarefoundation/auto_e2e#35]
## Context

The current BEV feature map resolution is 7×7, which comes directly from the backbone's final feature map size (256×256 input → 7×7 after 5 stages of downsampling). This was never intentionally designed as the BEV grid resolution — it's just the default backbone output that gets passed through.

At 7×7, if we assume a perception range of ~50m around the vehicle, each cell covers roughly 7m × 7m. A single passenger car (4.5m long) fits entirely within one cell, making it impossible to distinguish spatial relationships between nearby objects.

## How other models handle this

Most published BEV perception models define an explicit BEV grid independent of the backbone output resolution, then use attention or projection to populate it:

| Model | BEV Resolution | Perception Range | Cell Size | Reference |
|-------|---------------|-----------------|-----------|-----------|
| BEVFormer | 200×200 | 51.2m × 51.2m | 0.256m | [Li et al., ECCV 2022](https://arxiv.org/abs/2203.17270) |
| UniAD | 200×200 | 51.2m × 51.2m | 0.256m | [Hu et al., CVPR 2023](https://arxiv.org/abs/2212.10156) |
| BEVDet | 128×128 | 51.2m × 51.2m | 0.4m | [Huang et al., 2022](https://arxiv.org/abs/2112.11790) |
| BEVDepth | 128×128 | 51.2m × 51.2m | 0.4m | [Li et al., AAAI 2023](https://arxiv.org/abs/2206.10092) |
| PETR v2 | 100×100 | 51.2m × 51.2m | 0.512m | [Liu et al., ICCV 2023](https://arxiv.org/abs/2206.01256) |
| VAD | 200×200 | 60m × 30m | 0.3m / 0.15m | [Jiang et al., ICCV 2023](https://arxiv.org/abs/2303.12077) |
| SparseBEV | 200×200 | 51.2m × 51.2m | 0.256m | [Liu et al., ICCV 2023](https://arxiv.org/abs/2308.09244) |
| StreamPETR | 100×100 | 51.2m × 51.2m | 0.512m | [Wang et al., ICCV 2023](https://arxiv.org/abs/2303.11926) |

The common range is **100×100 to 200×200**, giving cell sizes of **0.25m to 0.5m**. This is fine enough to distinguish individual vehicles, lane markings, and pedestrians.

## Proposed Approach

Define an explicit BEV grid in the BEV fusion module, decoupled from backbone output:

```python
# Instead of using backbone's 7×7 directly:
bev_h, bev_w = 100, 100  # or 200×200
perception_range = 51.2  # meters in each direction

# Learnable BEV queries
bev_queries = nn.Parameter(torch.randn(bev_h * bev_w, embed_dim))

# Use cross-attention to populate from multi-view features
# (camera_params tells the attention where each BEV query maps to in each camera)
```

## Trade-offs to consider

| Resolution | Cell Size (50m range) | VRAM (embed_dim=256, fp16) | Compute |
|-----------|----------------------|---------------------------|---------|
| 50×50 | 1.0m | ~3 MB | Low |
| 100×100 | 0.5m | ~13 MB | Medium |
| 200×200 | 0.25m | ~50 MB | High |

Given that AutoE2E targets real-time inference on automotive hardware, **100×100 might be a reasonable starting point** — 0.5m resolution is enough for trajectory planning without the VRAM cost of 200×200.

## Questions

- What perception range should we target? 50m? 100m forward-biased (e.g., 80m front, 20m rear)?
- Should we start with 100×100 and see if it's sufficient for trajectory prediction, or go straight to 200×200?
- Does this interact with #23 (embed_dim unification)? A smaller embed_dim (256 vs 1440) would make higher BEV resolution more feasible.

Would appreciate thoughts on what makes sense for our use case and hardware targets.

**Comment by m-zain-khawaja** [comment]
Thanks @riita10069 
- I think forward bias would be important for safety critical actions, especially for higher speed driving where long-range perception is necessary and could be provided by narrow h-fov cameras such as the 30-degree h-fov front camera in the Nvidia Physical AI dataset. Although most approaches in the literature limit the BEV to be short/mid-range (since this is also where they are typically assessed for in benchmarks), I think it would be useful if we could try a BEV grid setup like:
    - 120m front, 60m rear, 60m left, 60m right
 - In this case, the BEV grid dimensions could be 450 x 300, which would give a cell size of 0.4m
 - Given the small embed dim of 256, this should hopefully be feasible without being too much more computationally expensive than 200 x 200 approaches.
 - My estimate for the type of hardware which would run this model would be something like Renesas RCar X5H and I think it should be able to run the model at 10Hz or faster.
 - This should hopefully make the model more practical and safe in real-world driving scenarios
- We could also try standard BEV grid size approaches:
   - a) 200 x 200 with 51.2m x 51.2m - similar to BEVFormer and UniAD
   - b) 200 x 200 with 60m x 30m - similar to VAD

**Comment by riita10069** [comment]
@m-zain-khawaja Thanks for the detailed response. The forward-biased 450×300 grid (120m front, 60m rear, 60m left/right, 0.4m cell) makes a lot of sense for safety-critical high-speed scenarios.

## Implementation Plan

To support this BEV resolution, the **DrivingPolicy (trajectory prediction head) needs to be redesigned**. The current approach of flattening the BEV feature map into a single vector won't scale — at 450×300×256, that would be 34.5M input dimensions to an MLP.

Looking at how this is handled in the literature:

| Model | BEV → Trajectory Method |
|-------|------------------------|
| UniAD (CVPR 2023) | Ego query + GRU autoregressive decoder. At each future timestep, the GRU hidden state cross-attends to BEV via **deformable attention**, then MLP decodes a waypoint. |
| VAD (ICCV 2023) | BEV → vectorized agent/map queries → ego query attends to these compact representations |
| BEVFormer (Waymo) | Already uses asymmetric grid (75m front / 35m rear = 300×220) with deformable attention |

The **ego query + deformable cross-attention** approach is well-suited for our asymmetric grid because:
- The ego query size is fixed (256-dim) regardless of BEV resolution
- Deformable attention scales linearly with BEV size (not quadratic like full attention)
- The model can learn to dynamically attend to different BEV regions based on driving context (e.g., look farther ahead at high speed)

## Proposed Architecture Change

Replace the current `DrivingPolicy` (flatten → MLP) with a `TrajectoryPlanner`:

```python
class TrajectoryPlanner(nn.Module):
    """Autoregressive trajectory decoder via ego query attending to BEV features."""

    def __init__(self, embed_dim=256, num_timesteps=64, num_signals=2):
        # Learnable ego query
        self.ego_query = nn.Embedding(1, embed_dim)
        # GRU for autoregressive decoding
        self.gru = nn.GRU(embed_dim, embed_dim, batch_first=True)
        # Deformable cross-attention to BEV
        self.cross_attn = DeformableCrossAttention(embed_dim, n_heads=8, n_points=4)
        # Egomotion conditioning
        self.ego_state_proj = nn.Linear(ego_dim, embed_dim)
        # Per-timestep waypoint output
        self.waypoint_head = nn.Linear(embed_dim, num_signals)

    def forward(self, bev_features, egomotion_history, bev_pos_embed):
        # bev_features: (B, 256, H, W) — works for any H×W
        # At each timestep: ego query cross-attends to BEV → GRU → waypoint
        ...
```

## Scope of Changes

| Component | Change |
|-----------|--------|
| `BEVViewFusion` | Configurable `bev_h`, `bev_w`, `pc_range` with asymmetric range support |
| `FeatureFusion` | Pass through variable BEV size (remove hardcoded 8×8 assumption) |
| `DrivingPolicy` → `TrajectoryPlanner` | Full rewrite: ego query + GRU + deformable cross-attention |
| `FutureState` | Accept variable BEV size (pool or attend) |
| Tests | Update expected shapes, add BEV resolution parameterization |

## Questions

1. Should the compressed visual feature vector (currently 14-dim from DrivingPolicy) still be produced? It's used by FutureState. With the new architecture, the ego query hidden state could serve the same purpose.
2. For the GRU horizon: 64 timesteps × 2 signals at 10Hz = 6.4s. Does this match the target for the 120m forward range at typical speeds? (At 120 km/h, 6.4s covers ~213m — sufficient.)
3. Should we keep the current `DrivingPolicy` as a fallback/legacy option, or replace it entirely?

Happy to take this on if the direction looks right.

**Comment by riita10069** [comment]
Following up on the implementation plan above — I'd like to proceed with the following decisions and submit a PR:

1. **Compressed visual vector (14-dim) → abolished.** The ego query hidden state (256-dim) from the GRU will be passed to FutureState instead. This is richer and more naturally couples trajectory planning with future prediction.

2. **DrivingPolicy → complete replacement** with the new `TrajectoryPlanner` (ego query + GRU + deformable cross-attention to BEV). No legacy fallback.

3. **Deformable attention → reuse existing `F.grid_sample` implementation** already in `bev_fusion.py`. The same reference points + learned offsets + bilinear sampling pattern applies to the trajectory planner attending to BEV features. No new dependencies needed.

@m-zain-khawaja would appreciate a quick confirmation on this direction before I open a PR. Specifically:
- Is removing the 14-dim compressed vector OK, or does anything else depend on it?
- Any concerns with full DrivingPolicy replacement?

**Comment by m-zain-khawaja** [comment]
Thanks @riita10069 - this sounds good to me and should improve the network considerably. The original 'Driving Policy' module was meant as a basic structure to build the foundation of the architecture but I think the proposed approach will position us much more closely to SOTA methods.

**Comment by m-zain-khawaja** [comment]
Note: @riita10069 - I am okay with abolishing the compressed visual feature vector, however, we would then no longer have any visual history input to the network (since the compressed visual feature vector was being put on a circular buffer to form the visual history input) and we would need to adjust the architecture of the Trajectory Planner to account for this. I think this change is fine, since the motivation of the visual history was to incorporate temporal information into the network. We can achieve this with other approaches such as temporal attention on the BEV, etc. in a separate PR.
